scp/gen_operators/SCPOperatorSysGen.cpp

```diff
@@ -0,0 +1,270 @@
+/*
+* This source file is part of an OSTIS project. For the latest info, see http://ostis.net
+* Distributed under the MIT License
+* (See accompanying file COPYING.MIT or copy at http://opensource.org/licenses/MIT)
+*/
+
+#include "scpUtils.hpp"
+#include "SCPOperatorSysGen.hpp"
+
+namespace scp
+{
+using namespace utils;
+SCPOperatorSysGen::SCPOperatorSysGen(const std::unique_ptr<ScMemoryContext> &ctx, ScAddr addr)
+  : SCPOperator(ctx, addr)
+{
+  operands = std::vector<SCPOperand*>(4);
+}
+
+std::string SCPOperatorSysGen::GetTypeName()
+{
+  return "sys_gen";
+}
+
+sc_result SCPOperatorSysGen::Parse()
+{
+  SCPOperator::Parse();
+  ScIterator3Ptr const operandIt3 = ms_context->Iterator3(addr, ScType::EdgeAccessConstPosPerm, ScType::Unknown);
+  while (operandIt3->Next())
+  {
+    auto * operand = new SCPOperand(ms_context, operandIt3->Get(1));
+
+    if (!(operand->GetOrder() > 0 && operand->GetOrder() < 5 && operands[operand->GetOrder() - 1] == nullptr))
+    {
+#ifdef SCP_DEBUG
+      std::stringstream ss;
+      ss << "Invalid operand order ";
+      ss << (int)operand->GetOrder();
+      Utils::logSCPError(ms_context, ss.str(), addr);
+#endif
+      FinishExecutionWithError();
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+    }
+    else
+    {
+      operands[operand->GetOrder() - 1] = operand;
+    }
+  }
+
+  for (size_t i = 0; i < 3; ++i)
+  {
+    auto * operand = operands[i];
+    if (!operand)
+    {
+#ifdef SCP_DEBUG
+      Utils::logSCPError(ms_context, "One or more operands missed", addr);
+      Utils::printInfo(ms_context, addr);
+#endif
+      FinishExecutionWithError();
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+    }
+  }
+  return SC_RESULT_OK;
+}
+
+sc_result SCPOperatorSysGen::Execute()
+{
+  if (operands[0]->IsAssign())
+  {
+#ifdef SCP_DEBUG
+    Utils::logSCPError(ms_context, "Operand 1 must have FIXED modifier", addr);
+#endif
+    FinishExecutionWithError();
+    return SC_RESULT_ERROR_INVALID_PARAMS;
+  }
+
+  if (operands[2]->IsAssign())
+  {
+#ifdef SCP_DEBUG
+    Utils::logSCPError(ms_context, "Operand 3 must have FIXED modifier", addr);
+#endif
+    FinishExecutionWithError();
+    return SC_RESULT_ERROR_INVALID_PARAMS;
+  }
+
+  if (operands[0]->IsFixed())
+  {
+    if (!operands[0]->GetValue().IsValid())
+    {
+#ifdef SCP_DEBUG
+      Utils::logSCPError(ms_context, "Operand 1 has modifier FIXED, but has no value", addr);
+#endif
+      FinishExecutionWithError();
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+    }
+  }
+
+  if (operands[2]->IsFixed())
+  {
+    if (!operands[2]->GetValue().IsValid())
+    {
+#ifdef SCP_DEBUG
+      Utils::logSCPError(ms_context, "Operand 3 has modifier FIXED, but has no value", addr);
+#endif
+      FinishExecutionWithError();
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+    }
+  }
```

Comment:
Can you move all this validation code to the separate method? And reuse it in sys search for example

Reply:
It is a problem for all operators in scp-machine